src/vuln_analysis/utils/java_segmenters_with_methods.py

```diff
@@ -0,0 +1,1220 @@
+                                if not sp or not (sp[0] <= pos < sp[1]):
+                                    pre = s[pos - 1] if pos > a else ''
+                                    post = s[pos + 3] if pos + 3 < b else ''
+                                    if (pos == a or not _is_ident_part(pre)) and (pos + 3 >= b or not _is_ident_part(post)):
+                                        return True
+                                pos = s.find('new', pos + 3, b)
+                            return False
+                        if not _has_new_between(src, left_bound, name_start):
+                            encl_reg = _enclosing_region(name_start)
+                            if not (encl_reg and name == encl_reg.name) and _prev_word(src, name_start) != 'record':
+                                body_end = _match_balanced_braces(src, i)
+                                sig_guess = _find_left_boundary(src, name_start)
+                                sig_start = _skip_leading_annotations(src, _skip_ws_comments(src, sig_guess))
+                                sig_start = _skip_ws_comments(src, sig_start)
+
+                                header_no_ann = _strip_annotations_in_range(src, sig_start, i)
+                                clean_header = _strip_comments_keep_strings_range(header_no_ann, 0, len(header_no_ann))
+
+                                clean_body = _strip_body_comments_and_annotations(src, i, body_end)
+
+                                method_text = clean_header + clean_body
+
+                                span = (sig_start, body_end)
+                                if span not in seen_spans:
+                                    seen_spans.add(span)
+                                    out.append(method_text)
+
+                                if i + 1 <= body_end and _body_might_have_local_types(i + 1, body_end):
+                                    out.extend(_extract_methods_anywhere(src, start=i + 1, end=body_end, types=types))
+                                i = body_end + 1
+                                continue
+        i += 1
+
+    return out
+
+# ============================== public API ====================================
+
+def extract_methods(java_source: str) -> List[str]:
+    """
+    Returns a list of Java method bodies/signatures (with bodies) plus all arrow-lambda
+    forms found in the file. Excludes constructors, comments/Javadoc, and strips
+    any method-level annotations that start before the signature token.
+    Order: methods first, then lambdas. Optimized for large files by pre-indexing
+    comments/strings and delimiter pairs (O(1) matching for (), {}).
+    """
+    _set_delim_index(java_source)
+    try:
+        # Extract
+        methods = _extract_methods_anywhere(java_source, 0, len(java_source))
+        # Lambdas in a separate pass across the whole file (so we don't double-count)
+        lambdas = _extract_lambdas_in_range(java_source, 0, len(java_source))
```

Comment:
@tmihalac Eventually you included lambdas in the feature? ( it seems that the segmenter parse them out from sources into documents)

Reply:
The lambdas need more work, currently the parameter types are not derived yet and its not handled yet in the search logic I've opened a new jira for it https://issues.redhat.com/browse/APPENG-3848